File: breadth_first_search/breadth_first_search.go
package main

import "fmt"

type Node struct {
	name     int
	children []*Node
}

func NewNode(name int) *Node {
	return &Node{name: name, children: []*Node{}}
}

func (n *Node) AddChild(name int) *Node {
	child := NewNode(name)
	n.children = append(n.children, child)
	return n
}

func (n *Node) BreadthFirstSearch(array *[]int) []int {
	queue := []*Node{n}
	for len(queue) > 0 {
		currentNode := queue[0]
		queue = queue[1:]
		*array = append(*array, currentNode.name)
		for _, child := range currentNode.children {
			queue = append(queue, child)
		}
	}
	return *array
}

func main() {
	myTree := NewNode(1).AddChild(2).AddChild(3).AddChild(4)
	result := myTree.BreadthFirstSearch(&[]int{})
	fmt.Println(result) // Output: [1 2 3 4]
}

File: breadth_first_search/breadth_first_search.ts
class MyNode {
  name: number;
  children: MyNode[];

  constructor(name: number) {
    this.name = name;
    this.children = [];
  }

  addChild(name: number): this {
    this.children.push(new MyNode(name));
    return this;
  }

  breadthFirstSearch(array: number[]): number[] {
    const queue: MyNode[] = [this];
    while (queue.length > 0) {
      const currentMyNode = queue.shift()!;
      array.push(currentMyNode.name);
      for (const child of currentMyNode.children) {
        queue.push(child);
      }
    }
    return array;
  }
}

const myTree = new MyNode(1).addChild(2).addChild(3).addChild(4);
console.log(myTree.breadthFirstSearch([])); // Output: [1, 2, 3, 4]
